### PR TITLE
parser: implement Restore for SelectStmt

### PR DESCRIPTION
**Pull request by AndrewDi**
implement Restore for `SelectStmt`
implement Restore for `UnionSelectList`
implement Restore for `SubqueryExpr`
implement Restore for `CompareSubqueryExpr`
implement Restore for `TableOptimizerHint`

Add `SelectStmt` test case, the rest would be add soon, too many test case.

Issue: https://github.com/pingcap/tidb/issues/8532

### CODE REVIEW

**Comment by AndrewDi** [comment]
@kennytm You can free to choose one, I implement it only because `SelectStmt` contains it.

**Comment by exialin** [comment]
@AndrewDi Since you've resolved the problem, it is better to include `TableOptimizerHint` in this PR. #151 can be closed later.

File: ast/dml.go [hunk 1]
```diff
@@ -755,7 +755,96 @@ type SelectStmt struct {
...
+
+	if n.Having != nil {
+		ctx.WritePlain(" ")
+		if err := n.Having.Restore(ctx); err != nil {
+			errors.Annotate(err, "An error occurred while restore SelectStmt.Having")
+		}
+	}
+
+	if n.WindowSpecs != nil {
+		//todo add Windows Func spec
```
Comment (exialin) [inline ast/dml.go]: `WindowSpecs` has already been merged.

Comment (AndrewDi) [reply ast/dml.go]: Done @exialin PTAL

Comment (AndrewDi) [reply ast/dml.go]: Test case has been added.

**Comment by AndrewDi** [comment]
@kennytm @leoppro Almost all the `SELECT` stmt I have found in `Parser_test` has been added, But it still have one restore problem here:
https://github.com/pingcap/parser/blob/d8e164fce4a69be70dd168ed470e88afb9b234f5/parser_test.go#L1430-L1447
If we restore a statement like `SELECT .78`, output would become `SELECT 0.78`, so the `MyDeciaml`'s `digitsInt` will change from `0` to `1`, `DeepEquals` will fail.

Besides `SELECT WITH Exists Clause` test case should add after #126 merged.

**Comment by leoppro** [comment]
> @kennytm @leoppro Almost all the `SELECT` statement test case I have found in `Parser_test` has been added, But it still have one restore problem here:
> [parser/parser_test.go](https://github.com/pingcap/parser/blob/d8e164fce4a69be70dd168ed470e88afb9b234f5/parser_test.go#L1430-L1447)
> 
> Lines 1430 to 1447 in [d8e164f](/pingcap/parser/commit/d8e164fce4a69be70dd168ed470e88afb9b234f5)
> 
>  {`select .78+123`, true, ""}, 
>  {`select .78+.21`, true, ""}, 
>  {`select .78-123`, true, ""}, 
>  {`select .78-.21`, true, ""}, 
>  {`select .78--123`, true, ""}, 
>  {`select .78*123`, true, ""}, 
>  {`select .78*.21`, true, ""}, 
>  {`select .78/123`, true, ""}, 
>  {`select .78/.21`, true, ""}, 
>  {`select .78,123`, true, ""}, 
>  {`select .78,.21`, true, ""}, 
>  {`select .78 , 123`, true, ""}, 
>  {`select .78.123`, false, ""}, 
>  {`select .78#123`, true, ""}, // select .78 
>  {`insert float_test values(.67, 'string');`, true, ""}, 
>  {`select .78'123'`, true, ""}, // select .78 as '123' 
>  {"select .78`123`", true, ""}, // select .78 as `123` 
>  {`select .78"123"`, true, ""}, // select .78 as "123" 
> 
> If we restore a statement like `SELECT .78`, output would become `SELECT 0.78`, so the `MyDeciaml`'s `digitsInt` will change from `0` to `1`, `DeepEquals` will fail.
> Besides `SELECT WITH Exists Clause` test case should add after #126 merged.

could you handle this by updating `CleanNodeText`

**Comment by AndrewDi** [comment]
this attr could update now ...

**Comment by AndrewDi** [comment]
@leoppro PTAL, use another way to deal with `Decimal` type.

File: ast/dml.go [hunk 2]
```diff
@@ -756,7 +756,110 @@ type SelectStmt struct {
...
+
+	if !n.SelectStmtOpts.SQLCache {
+		ctx.WriteKeyWord("SQL_NO_CACHE ")
+	}
+
+	if n.TableHints != nil && len(n.TableHints) != 0 {
+		ctx.WritePlain("/*+ ")
+		for _, tableHint := range n.TableHints {
+			if err := tableHint.Restore(ctx); err != nil {
+				errors.Annotate(err, "An error occurred while restore SelectStmt.TableHints")
```
Comment (leoppro) [inline ast/dml.go]:
```suggestion
				errors.Annotatef(err, "An error occurred while restore SelectStmt.TableHints[%d]", i)
```

File: ast/dml.go [hunk 3]
```diff
@@ -756,7 +756,110 @@ type SelectStmt struct {
...
+	if n.SelectStmtOpts.StraightJoin {
+		ctx.WriteKeyWord("STRAIGHT_JOIN ")
+	}
+	if n.Fields != nil {
+		for i, field := range n.Fields.Fields {
+			if i != 0 {
+				ctx.WritePlain(",")
+			}
+			if err := field.Restore(ctx); err != nil {
+				errors.Annotate(err, "An error occurred while restore SelectStmt.Field")
```
Comment (leoppro) [inline ast/dml.go]: ditto

File: ast/dml.go [hunk 4]
```diff
@@ -756,7 +756,110 @@ type SelectStmt struct {
...
+	}
+
+	if n.WindowSpecs != nil {
+		ctx.WriteKeyWord(" WINDOW ")
+		for i, windowsSpec := range n.WindowSpecs {
+			if i != 0 {
+				ctx.WritePlain(",")
+			}
+			if err := windowsSpec.Restore(ctx); err != nil {
+				errors.Annotate(err, "An error occurred while restore SelectStmt.WindowSpec")
```
Comment (leoppro) [inline ast/dml.go]: ditto

File: ast/dml.go [hunk 5]
```diff
@@ -756,7 +756,110 @@ type SelectStmt struct {
...
+	}
+
+	if n.Limit != nil {
+		ctx.WritePlain(" ")
+		if err := n.Limit.Restore(ctx); err != nil {
+			errors.Annotate(err, "An error occurred while restore SelectStmt.Limit")
+		}
+	}
+
+	switch n.LockTp {
```
Comment (leoppro) [inline ast/dml.go]: use `if` or add `default`

File: ast/functions.go [hunk 6]
```diff
@@ -644,12 +646,28 @@ func (n *AggregateFuncExpr) Restore(ctx *RestoreCtx) error {
...
-		if i != 0 {
-			ctx.WritePlain(", ")
+	switch strings.ToLower(n.F) {
+	case "group_concat":
+		for i := 0; i < len(n.Args)-1; i++ {
+			if i != 0 {
+				ctx.WritePlain(", ")
+			}
+			if err := n.Args[i].Restore(ctx); err != nil {
+				return errors.Annotatef(err, "An error occurred while restore AggregateFuncExpr.Args %d", i)
```
Comment (leoppro) [inline ast/functions.go]:
```suggestion
				return errors.Annotatef(err, "An error occurred while restore AggregateFuncExpr.Args[%d]", i)
```

File: ast/functions.go [hunk 7]
```diff
@@ -644,12 +646,28 @@ func (n *AggregateFuncExpr) Restore(ctx *RestoreCtx) error {
...
+			}
+		}
+		ctx.WriteKeyWord(" SEPARATOR ")
+		if err := n.Args[len(n.Args)-1].Restore(ctx); err != nil {
+			return errors.Annotatef(err, "An error occurred while restore AggregateFuncExpr.Args SEPARATOR")
```
Comment (leoppro) [inline ast/functions.go]:
```suggestion
			return errors.Annotate(err, "An error occurred while restore AggregateFuncExpr.Args SEPARATOR")
```

File: ast/functions.go [hunk 8]
```diff
@@ -644,12 +646,28 @@ func (n *AggregateFuncExpr) Restore(ctx *RestoreCtx) error {
...
 		}
-		if err := argv.Restore(ctx); err != nil {
-			return errors.Annotatef(err, "An error occurred while restore AggregateFuncExpr.Args %d", i)
+	default:
+		for i, argv := range n.Args {
+			if i != 0 {
+				ctx.WritePlain(", ")
+			}
+			if err := argv.Restore(ctx); err != nil {
+				return errors.Annotatef(err, "An error occurred while restore AggregateFuncExpr.Args %d", i)
```
Comment (leoppro) [inline ast/functions.go]: ditto

File: ast/dml.go (hunk 5 above)
Comment (AndrewDi) [reply ast/dml.go]: By default, locktp is should be empty

File: ast/expressions.go [hunk 9]
```diff
@@ -817,6 +836,12 @@ func (n *PatternLikeExpr) Restore(ctx *RestoreCtx) error {
 		return errors.Annotate(err, "An error occurred while restore PatternLikeExpr.Pattern")
 	}
 
+	escape := string(n.Escape)
+	if escape != "\\" {
+		ctx.WriteKeyWord(" ESCAPE ")
+		ctx.WritePlainf("'%s'", escape)
```
Comment (kennytm) [inline ast/expressions.go]:
```suggestion
		ctx.WriteString(escape)
```

Demo:

```sql
select '''_' like '''_' escape '''';
```

File: ast/misc.go [hunk 10]
```diff
@@ -1045,7 +1045,20 @@ type TableOptimizerHint struct {
 
 // Restore implements Node interface.
 func (n *TableOptimizerHint) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	ctx.WriteKeyWord(n.HintName.String())
+	ctx.WritePlain("(")
+	if n.HintName.O == "MAX_EXECUTION_TIME" {
```
Comment (kennytm) [inline ast/misc.go]:
```suggestion
	if n.HintName.L == "max_execution_time" {
```

I suppose `/*+ max_execution_time(3000) */` is a valid hint.

**Comment by AndrewDi** [comment]
@kennytm Comment handled

**Pull request merged**